deployUtils.ts

```diff
@@ -0,0 +1,122 @@
+import hre, { ethers } from 'hardhat';
+import fs from 'fs';
+import { Contract } from 'ethers';
+const OUTPUT_FILE = 'output.json';
+
+export type DeploymentAddresses = {
+  walletImplementation?: string;
+  walletFactory?: string;
+  forwarderImplementation?: string;
+  forwarderFactory?: string;
+};
+/**
+ * Attempts to load contract address from output file or by predicting nonce.
+ * If neither found, it deploys the contract using the given deploy function.
+ */
+// Core reusable function
+export async function deployIfNeededAtNonce(
+  recordedAddress: string | undefined,
+  expectedNonce: number,
+  deployerAddress: string,
+  contractName: string,
+  deployFn: () => Promise<string>
+): Promise<string> {
+  const predictedAddress = ethers.utils.getContractAddress({
+    from: deployerAddress,
+    nonce: expectedNonce
+  });
+
+  console.log(
+    `🔮 Expecting ${contractName} at nonce ${expectedNonce} -> ${predictedAddress}`
+  );
+
+  // 1. Check output.json record
+  if (recordedAddress) {
```

Comment:
since the output file is created during the process
I don't get when the recordedAddress will ever be defined